pkg/cable/wireguard/driver.go

```diff
@@ -0,0 +1,411 @@
+
+	if keyMismatch {
+		// ClusterID probably already associated with new key. Do not remove peers entry nor routes
+		klog.Warningf("Key mismatch for peer cluster %s, keeping existing routes and public key", remoteEndpoint.Spec.ClusterID)
+		return nil
+	}
+	// no delete on mismatch
+	delete(w.peers, remoteEndpoint.Spec.ClusterID)
+
+	// del routes
+	allowedIPs := parseSubnets(remoteEndpoint.Spec.Subnets)
+	idx := w.link.Attrs().Index
+	for _, peerNet := range allowedIPs {
+		route := netlink.Route{
+			LinkIndex: idx,
+			Dst:       &peerNet,
+		}
+		if err = netlink.RouteDel(&route); err != nil {
+			return fmt.Errorf("failed to delete route %s: %v", route, err)
+		}
+	}
+
+	klog.V(log.DEBUG).Infof("Done removing endpoint for cluster %s", remoteEndpoint.Spec.ClusterID)
+	return nil
+}
+
+func (w *wireguard) GetActiveConnections(clusterID string) ([]string, error) {
+	// force caller to skip duplicate handling
+	return make([]string, 0), nil
+}
+
+// Create new wg link and assign addr from local subnets
+func setWGLink(w *wireguard, localSubnets []string) error {
+
+	// delete existing wg device if needed
+	if link, err := netlink.LinkByName(DefaultDeviceName); err == nil {
+		// delete existing device
+		if err := netlink.LinkDel(link); err != nil {
+			return fmt.Errorf("failed to delete existing WireGuard device: %v", err)
+		}
+	}
+
+	// create the wg device (ip link add dev $DefaultDeviceName type wireguard)
+	la := netlink.NewLinkAttrs()
+	la.Name = DefaultDeviceName
+	link := &netlink.GenericLink{
+		LinkAttrs: la,
+		LinkType:  "wireguard",
+	}
+	if err := netlink.LinkAdd(link); err == nil {
+		w.link = link
+	} else {
+		return fmt.Errorf("failed to add WireGuard device: %v", err)
+	}
+
+	// parse localSubnets and get internal address
+	w.localSubnets = parseSubnets(localSubnets)
+	ip, err := discoverInternalIP(w.localSubnets)
+	if err != nil {
+		klog.V(log.DEBUG).Infof("Using endpoint IP as internal address; %v", err)
+		ip = endpointIP(&w.localEndpoint)
+	}
+	klog.V(log.DEBUG).Infof("Setting interface address to  %s", ip)
+
+	// setup local address (ip address add dev $DefaultDeviceName $PublicIP
+	localIP, err := netlink.ParseAddr(ip + "/32")
+	if err != nil {
+		// try again as CIDR
+		if localIP, err = netlink.ParseAddr(ip); err != nil {
+			return fmt.Errorf("failed to parse IP address %s: %v", ip, err)
+		}
+	}
+	if err = netlink.AddrAdd(w.link, localIP); err != nil {
+		return fmt.Errorf("failed to add local address: %v", err)
+	}
+
+	return nil
+}
+
+// find internal host IP inside one of the local CIDRs
+// TODO move this to utils
+func discoverInternalIP(cidrs []net.IPNet) (string, error) {
+
+	addrs, err := net.InterfaceAddrs()
+	if err != nil {
+		return "", fmt.Errorf("net.InterfaceAddrs() returned error : %v", err)
+	}
+
+	for _, a := range addrs {
+		ip, _, err := net.ParseCIDR(a.String())
+		if err != nil {
+			klog.V(log.DEBUG).Infof("Skipping local address %v, unable to ParseCIDR: %v", a, err)
```

Comment:
It seems this should be logged as error or warning as we do on L365? I assume this isn't expected...

Reply:
I don't have a strong opinion either way, but the errors are slightly different. 
L365 is submariner error -- malformed subnets -- we can continue but something is probably not working as sub CIDRs will not be added to allowed IPs and not routed to other cluster.
L342 is netlink library error or some weird device that returns an unexpected address -- we simply don't use it and look for another internal address. Only if we cannot find any address this becomes an error at L355.

Reply:
👍 makes sense.